lib/util/LocationStatusManager.js

```diff
@@ -59,6 +195,452 @@ class LocationStatusManager {
+     */
+    async _getZkDataForLocation(service, location) {
+        const zkNamespace = this._serviceConfig[service]?.namespace;
+        const zkStatePath = this._serviceConfig[service]?.statePath;
+
+        if (!zkNamespace || !zkStatePath) {
+            this._logger.error('ZooKeeper namespace or statePath not configured for service',
+                { service, location });
+            return { paused: false, scheduledResume: null }; // Default to running if config is missing
+        }
+
+        const path = `${zkNamespace}${zkStatePath}/${location}`;
+        return new Promise((resolve, reject) => {
+            this._zkClient.getData(path, (err, data) => {
+                if (err) {
+                    if (err.name === 'NO_NODE') {
+                        this._logger.info('ZooKeeper node not found, assuming default (not paused) state',
+                            { path, service, location });
+                        return resolve({ paused: false, scheduledResume: null });
+                    }
+                    this._logger.error('Error getting ZooKeeper data for location',
+                        { path, service, location, error: err.message });
+                    return reject(err);
+                }
+                try {
+                    const parsedData = JSON.parse(data.toString());
+                    this._logger.debug('Successfully fetched ZooKeeper data for location',
+                        { path, service, location, data: parsedData });
+                    return resolve(parsedData);
+                } catch (e) {
+                    this._logger.error('Error parsing ZooKeeper data for location',
+                        { path, service, location, rawData: data.toString(), error: e.message });
+                    return reject(e); 
+                }
+            });
+        });
+    }
+
+
+    /**
+     * Get/Create state machine for service-location pair
+     * @param {string} service service name
+     * @param {string} location location name
+     * @return {Promise<ServiceLocationStateMachine>} state machine instance
+     */
+    async _getStateMachine(service, location) {
+        const key = `${service}:${location}`;
+        // eslint-disable-next-line no-console
+        console.log('checking state machine for key', this._stateMachines.has(key));
+        if (!this._stateMachines.has(key)) {
+            let initialState = SERVICE_STATES.RUNNING;
+            let sourceData = null; 
+            const serviceConfig = this._serviceConfig[service];
+            this._logger.debug('Service Config state machine', { serviceConfig });
+            if (this._serviceConfig[service] && !this._serviceConfig[service].isMongo) {
+                this._logger.debug('Initializing state machine for ZK-based service', { key, service, location });
+                try {
+                    sourceData = await this._getZkDataForLocation(service, location);
```

Comment:
This calls ZK every time we process an operation, is that OK? I saw we increased the timeouts in tests from 1000ms to 2000ms, is this change fine if we pause/resume at high rate?